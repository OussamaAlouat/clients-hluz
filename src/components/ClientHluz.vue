<template>
  <v-card class="mx-auto" max-width="344" outlined>
    <v-list-item three-line>
      <v-list-item-content>
        <v-list-item-title class="text-h5 mb-1">
          Cecil Grant
        </v-list-item-title>
        <v-list-item-subtitle>Sweet potato street, 2</v-list-item-subtitle>
        <v-list-text>customer-basic</v-list-text>
      </v-list-item-content>

      <v-list-item-avatar tile size="80" color="grey">Build</v-list-item-avatar>
    </v-list-item>
  </v-card>
</template>

<script>

export default {
  name: 'ClientHluz',
  props: {
    clientName: {
      type: String,
      required: true,
    },

    adress: {
      type: String,
      required: true,
    },

    role: {
      type: String,
      required: true,
    },
    buildingType: {
      type: String,
      required: true,
    },
  },
};
</script>

<style>

</style>
