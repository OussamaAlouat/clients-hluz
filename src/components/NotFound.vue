<template>
  <div>
    <v-alert type="info" icon="mdi-home-alert">
      There are no clients for that CUT
    </v-alert>
  </div>
</template>

<script>
export default {
  name: 'NotFound',
};
</script>

<style lang="scss">
</style>
