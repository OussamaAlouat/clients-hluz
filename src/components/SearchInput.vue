<!-- eslint-disable vuejs-accessibility/form-control-has-label -->
<template>
  <div  class="input-container">
    <div style="width:80%">
      <v-text-field
        placeholder="Set cups"
        outlined
        v-model="cupsValue"
        class="input-container__input-hluz"
        type="text"
        id="cups"
      >
        <template v-slot:append>
          <v-hover v-slot="{ hover }">
            <v-icon
              large
              :color="hover ? 'rgba(229, 0, 126, 255)' : ''"
              :disabled="!cupsValue" @click="searchCUPS()"
            >
              mdi-magnify
            </v-icon>
          </v-hover>
        </template>
      </v-text-field>
    </div>

  </div>
</template>

<script>
export default {
  name: 'SearchInput',
  data: () => ({
    cupsValue: '',
  }),

  methods: {
    searchCUPS() {
      if (this.cupsValue) {
        this.$emit('cupsToSearch', this.cupsValue);
      }
    },
  },
};
</script>

<style lang="scss">
.input-container {
  display: flex;
  justify-content: center;
  margin-top: 20px;
  margin-bottom: 20px;

  &__input-hluz {
    border-radius: 10px;
    height: 60px;
    justify-content: center;
  }
}
</style>
