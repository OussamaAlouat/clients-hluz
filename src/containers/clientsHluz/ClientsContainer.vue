<template>
  <div class="h-100">
    <search-input @cupsToSearch="setCupsToSearch"></search-input>
    <div
      class="d-flex justify-center"
      v-if="clientsRooftopRevolution && clientsRooftopRevolution.full_name"
    >
      <client-hluz
        :clientName="clientsRooftopRevolution.full_name"
        :address="clientsRooftopRevolution.address"
        :role="clientsRooftopRevolution.role"
        :buildingType="clientsRooftopRevolution.building_type"
      >
      </client-hluz>
    </div>

    <div class="offer-container">
      <offer-hluz
        v-for="(amount, index) in amounts"
        :key="index"
        :amount="amount.amount"
        :type="amount.type"
        :icon="amount.icon"
      ></offer-hluz>
    </div>

    <div v-if="showNotFoundClients" class="h-100 not-found-container">
      <not-found />
    </div>
  </div>
</template>

<script>
import api from '@/services/api';

import ClientHluz from '@/components/ClientHluz.vue';
import SearchInput from '@/components/SearchInput.vue';
import NotFound from '@/components/NotFound.vue';
import OfferHluz from '@/components/OfferHluz.vue';

export default {
  components: {
    ClientHluz,
    SearchInput,
    NotFound,
    OfferHluz,
  },

  data: () => ({
    clientsWithCuts: {},
    supplyPoint: {},
    searched: false,
    neighborsInfo: [],
    amounts: [],
  }),
  mounted() {
    api.getAllClients().then((result) => {
      console.log(result.data);
    });
  },
  methods: {
    resetData() {
      this.supplyPoint = {};
      this.neighborsInfo = [];
      this.amounts = [];
      this.clientsWithCuts = {};
    },

    async setCupsToSearch(event) {
      this.resetData();
      const clients = api.getClientsWithCups(event);
      const supplyPointPromise = api.getSupplyPoints(event);
      Promise.all([clients, supplyPointPromise]).then((values) => {
        // eslint-disable-next-line prefer-destructuring
        this.clientsWithCuts = values[0].data[0];
        const resultOfSuplyPromise = values[1];

        if (resultOfSuplyPromise.data && resultOfSuplyPromise.data.length > 0) {
        // eslint-disable-next-line prefer-destructuring
          this.supplyPoint = resultOfSuplyPromise.data[0];
          this.getAllNeighborsInformation();
        }

        this.searched = true;
      });
    },

    getAllNeighborsInformation() {
      if (this.supplyPoint && this.supplyPoint.neighbors) {
        this.supplyPoint.neighbors.forEach(async (elem) => {
          const value = await this.getNeighborsInformation(elem);
          this.neighborsInfo.push(...value);
        });
      }
    },

    async getNeighborsInformation(neighborCups) {
      const neighbor = await api.getSupplyPoints(neighborCups);
      return neighbor.data;
    },
  },
  computed: {
    numberOfNeighbors() {
      if (this.supplyPoint && this.supplyPoint.neighbors) {
        return this.supplyPoint?.neighbors.length;
      }

      return 0;
    },

    clientsRooftopRevolution() {
      if (this.clientsWithCuts && this.clientsWithCuts.building_type === 'house' && this.numberOfNeighbors >= 1) {
        return this.clientsWithCuts;
      }

      return {};
    },

    showNotFoundClients() {
      return (
        this.clientsRooftopRevolution
        && !this.clientsRooftopRevolution.full_name
        && this.searched
      );
    },

    isClientHaveBasicDiscount() {
      // eslint-disable-next-line no-plusplus
      for (let i = 0; i < this.neighborsInfo.length; i++) {
        const { p1, p2 } = this.neighborsInfo[i].power;

        if (p1 >= this.supplyPoint.power.p1 || p2 >= this.supplyPoint.power.p2) {
          return false;
        }
      }

      if (this.neighborsInfo.length === 0) {
        return false;
      }

      return true;
    },

    isClientHaveSpecialDiscount() {
      let sum = 0;
      // eslint-disable-next-line no-plusplus
      for (let i = 0; i < this.neighborsInfo.length; i++) {
        console.log('this.neighborsInfo.length: ', this.neighborsInfo[i]);
        sum += parseFloat(this.neighborsInfo[i].invoiced_amount);
      }

      if (sum <= 100) {
        return false;
      }

      return true;
    },
  },

  watch: {
    neighborsInfo() {
      if (
        this.supplyPoint
        && this.supplyPoint.neighbors
        && this.neighborsInfo.length === this.supplyPoint.neighbors.length
      ) {
        const amount = this.supplyPoint.invoiced_amount;
        if (this.clientsRooftopRevolution && this.clientsRooftopRevolution.full_name) {
          this.amounts.push({
            type: 'Standard offer',
            amount,
            icon: 'mdi-lightbulb-on-10',
          });
        }

        if (this.isClientHaveBasicDiscount) {
          const amountBasic = parseFloat(amount) * 0.95;
          this.amounts.push({
            type: 'Basic discount',
            amount: amountBasic,
            icon: 'mdi-lightbulb-on-40',
          });
        }

        if (this.isClientHaveSpecialDiscount) {
          const amountSpecial = parseFloat(amount) * 0.88;
          this.amounts.push({
            type: 'Special discount',
            amount: amountSpecial,
            icon: 'mdi-lightbulb-on-90',
          });
        }
      }
    },
  },
};
</script>

<style lang="scss">
.h-100 {
  height: calc(100% - 60px);
}

.not-found-container{
  display: flex;
  justify-content: center;
  align-content: center;
  align-items: center;
}

.offer-container {
  display: flex;
  justify-content: space-evenly;
  margin-top: 40px;
}
</style>
