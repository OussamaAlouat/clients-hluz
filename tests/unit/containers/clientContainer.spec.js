import { shallowMount } from '@vue/test-utils';
import ClientsContainer from '@/containers/clientsHluz/ClientsContainer.vue';
import ClientHluz from '@/components/ClientHluz.vue';
import SearchInput from '@/components/SearchInput.vue';
import NotFound from '@/components/NotFound.vue';
import mocks from '../../__mocks__/mocks';

describe('ClientsContainer.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(ClientsContainer, {
      components: {
        ClientHluz,
        SearchInput,
        NotFound,
      },
    });
  });

  it('Container render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Displays not found when not found clients', async () => {
    wrapper.setData({ searched: true });
    await wrapper.vm.$nextTick();
    const notFound = wrapper.findComponent(NotFound);
    expect(notFound.exists()).toBe(true);
  });

  it('Displays one client card', async () => {
    wrapper = shallowMount(ClientsContainer, {
      components: {
        ClientHluz,
        NotFound,
        SearchInput,
      },
      computed: {
        showNotFoundClients: () => false,
        numberOfNeighbors: () => 2,
        clientsRooftopRevolution: () => mocks.curlClient,
      },
      data() {
        return {
          clientsWithCuts: mocks.curlClient,
          supplyPoint: mocks.selecedSuplyPoint,
          neighborsInfo: [
            ...mocks.suplyPoints,
          ],
          amounts: [],
          searched: true,
          loading: false,
        };
      },
    });

    await wrapper.vm.$nextTick();

    const clientHluz = wrapper.findComponent(ClientHluz);
    expect(clientHluz.exists()).toBe(true);
  });
});
