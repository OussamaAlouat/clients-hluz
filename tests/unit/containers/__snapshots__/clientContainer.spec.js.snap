// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientsContainer.vue Container render 1`] = `
<div class="h-100">
  <search-input-stub></search-input-stub>
  <!---->
</div>
`;
